momentGW/thc.py

```diff
@@ -0,0 +1,270 @@
+import numpy as np
+from h5py import File
+from pyscf import lib
+from pyscf.agf2 import mpi_helper
+from scipy.special import binom
+
+
+class Integrals:
+    """
+    Container for the integrals required for GW methods.
+    """
+
+    def __init__(
+        self,
+        with_df,
+        mo_coeff,
+        mo_occ,
+        thc_opts,
+    ):
+        self.with_df = with_df
+        self.mo_coeff = mo_coeff
+        self.mo_occ = mo_occ
+        self.filepath = thc_opts["file_path"]
+
+        self._blocks = {}
+
+    def transform(self):
+        """
+        Imports a H5PY file containing a dictionary. Inside the dict, a 'collocation_matrix' and
+        a 'coulomb_matrix' must be contained with shapes (aux, MO) and (aux,aux) respectively.
+        """
+        if self.filepath is None:
+            raise ValueError("filepath cannot be None for THC implementation")
+        thc_eri = File(self.filepath, "r")
+        coll = np.array(thc_eri["collocation_matrix"]).T[0].T
+        cou = np.array(thc_eri["coulomb_matrix"][0]).T[0].T
+        Xip = coll[: self.nocc, :]
+        Xap = coll[self.nocc :, :]
+        self._blocks["coll"] = coll
+        self._blocks["cou"] = cou
+        self._blocks["Xip"] = Xip
+        self._blocks["Xap"] = Xap
+
+    @property
+    def Coll(self):
+        """
+        Return the (aux, MO) array.
+        """
+        return self._blocks["coll"]
+
+    @property
+    def Cou(self):
+        """
+        Return the (aux, aux) array.
+        """
+        return self._blocks["cou"]
+
+    @property
+    def Xip(self):
+        """
+        Return the (aux, W occ) array.
+        """
+        return self._blocks["Xip"]
+
+    @property
+    def Xap(self):
+        """
+        Return the (aux, W vir) array.
+        """
+        return self._blocks["Xap"]
+
+    @property
+    def nmo(self):
+        """
+        Return the number of MOs.
+        """
+        return self.mo_coeff.shape[-1]
+
+    @property
+    def nocc(self):
+        """
+        Return the number of occupied MOs.
+        """
+        return np.sum(self.mo_occ > 0)
+
+    @property
+    def naux(self):
+        """
+        Return the number of auxiliary basis functions, after the
+        compression.
+        """
+        return self.Cou.shape[0]
+
+
+class TDA:
+    """
+    Compute the self-energy moments using THC integrals in TDA
+
+    Parameters
+    ----------
+    tda: TDA
+        TDA object
+    """
+
+    def __init__(
+        self,
+        gw,
+        nmom_max,
+        integrals,
+        mo_energy=None,
+        mo_occ=None,
+    ):
+        self.gw = gw
+        self.integrals = integrals
+        self.naux = self.integrals.naux
+        self.nmo = self.integrals.nmo
+        self.nmom_max = nmom_max
+        self.total_nmom = self.nmom_max + 1
+        self.nocc = self.integrals.nocc
+
+        self.XiP = self.integrals.Xip
+        self.XaP = self.integrals.Xap
+        self.Z = self.integrals.Cou
```

Comment:
as above, these should either be properties:
```
@property
def XiP(self):
    return self.integrals.Xip
```
or just not assigned and accessed as `self.integrals.Xip` throughout the class functions. I'd prefer the latter but if it makes lines too long and code ugly then feel free to add them as properties.